src/Http/Middleware/RateLimitMiddleware.php

```diff
@@ -0,0 +1,289 @@
+use Cake\Http\RateLimit\TokenBucketRateLimiter;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+/**
+ * Rate limiting middleware
+ *
+ * Provides configurable rate limiting based on various identifiers.
+ * Supports multiple strategies including sliding window, token bucket, and fixed window.
+ */
+class RateLimitMiddleware implements MiddlewareInterface
+{
+    /**
+     * Default configuration
+     *
+     * @var array<string, mixed>
+     */
+    protected array $defaultConfig = [
+        'limit' => 60,
+        'window' => 60,
+        'identifier' => 'ip',
+        'strategy' => 'sliding_window',
+        'cache' => 'default',
+        'headers' => true,
+        'message' => 'Rate limit exceeded. Please try again later.',
+        'skipCheck' => null,
+        'costCallback' => null,
+        'identifierCallback' => null,
+        'limitCallback' => null,
+    ];
+
+    /**
+     * Configuration
+     *
+     * @var array<string, mixed>
+     */
+    protected array $config;
+
+    /**
+     * Constructor
+     *
+     * @param array<string, mixed> $config Configuration options
+     */
+    public function __construct(array $config = [])
+    {
+        $this->config = $config + $this->defaultConfig;
+    }
+
+    /**
+     * Process the request and add rate limiting
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @param \Psr\Http\Server\RequestHandlerInterface $handler The handler
+     * @return \Psr\Http\Message\ResponseInterface
+     */
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        if ($this->shouldSkip($request)) {
+            return $handler->handle($request);
+        }
+
+        $identifier = $this->getIdentifier($request);
+        $limit = $this->getLimit($request, $identifier);
+        $window = $this->config['window'];
+        $cost = $this->getCost($request);
+
+        $rateLimiter = $this->getRateLimiter();
+        $result = $rateLimiter->attempt($identifier, $limit, $window, $cost);
+
+        if (!$result['allowed']) {
+            throw new TooManyRequestsException($this->config['message']);
+        }
+
+        $response = $handler->handle($request);
+
+        if ($this->config['headers']) {
+            $response = $this->addRateLimitHeaders($response, $result);
+        }
+
+        return $response;
+    }
+
+    /**
+     * Check if rate limiting should be skipped for this request
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @return bool
+     */
+    protected function shouldSkip(ServerRequestInterface $request): bool
+    {
+        if ($this->config['skipCheck'] === null) {
+            return false;
+        }
+
+        return (bool)call_user_func($this->config['skipCheck'], $request);
```

Comment:
Do we need to support string callables? We have been using `Closure` more which doesn't require `call_user_func`.